File: CPP/Problem_054.cpp
#include <vector>
#include <algorithm>
#include <unordered_map>
#include <fstream>
#include <iostream>

using namespace std;

static int rval(char c) {
    if (c >= '2' && c <= '9') return c - '0';
    if (c == 'T') return 10;
    if (c == 'J') return 11;
    if (c == 'Q') return 12;
    if (c == 'K') return 13;
    return 14; // A
}

// Score is a vector<int> where lexicographic compare works.
static vector<int> score(const vector<string>& hand) {
    vector<int> ranks(5);
    vector<char> suits(5);
    for (int i = 0; i < 5; i++) {
        ranks[i] = rval(hand[i][0]);
        suits[i] = hand[i][1];
    }
    sort(ranks.begin(), ranks.end());

    bool isFlush = all_of(suits.begin(), suits.end(), [&](char s){ return s == suits[0]; });

    bool isStraight = false;
    int straightHigh = ranks[4];
    bool consecutive = true;
    for (int i = 1; i < 5; i++) {
        if (ranks[i] != ranks[i-1] + 1) { consecutive = false; break; }
    }
    if (consecutive) {
        isStraight = true;
    } else if (ranks == vector<int>({2,3,4,5,14})) {
        isStraight = true;
        straightHigh = 5;
    }

    unordered_map<int,int> freq;
    for (int r : ranks) freq[r]++;

    vector<pair<int,int>> groups; // (count, rank)
    groups.reserve(freq.size());
    for (auto &kv : freq) groups.push_back({kv.second, kv.first});
    sort(groups.begin(), groups.end(), [](auto &a, auto &b){
        if (a.first != b.first) return a.first > b.first;
        return a.second > b.second;
    });

    vector<int> ranksDesc = ranks;
    reverse(ranksDesc.begin(), ranksDesc.end());

    if (isFlush && isStraight) {
        if (straightHigh == 14) return {9};            // royal flush
        return {8, straightHigh};                      // straight flush
    }

    if (groups[0].first == 4) {
        int quad = groups[0].second;
        int kicker = 0;
        for (int r : ranks) if (r != quad) kicker = max(kicker, r);
        return {7, quad, kicker};
    }

    if (groups[0].first == 3 && groups.size() > 1 && groups[1].first == 2) {
        return {6, groups[0].second, groups[1].second}; // full house
    }

    if (isFlush) {
        vector<int> v = {5};
        v.insert(v.end(), ranksDesc.begin(), ranksDesc.end());
        return v;
    }

    if (isStraight) return {4, straightHigh};

    if (groups[0].first == 3) {
        int trip = groups[0].second;
        vector<int> kickers;
        for (int r : ranks) if (r != trip) kickers.push_back(r);
        sort(kickers.rbegin(), kickers.rend());
        return {3, trip, kickers[0], kickers[1]};
    }

    if (groups[0].first == 2 && groups.size() > 1 && groups[1].first == 2) {
        int p1 = groups[0].second, p2 = groups[1].second;
        int highPair = max(p1, p2), lowPair = min(p1, p2);
        int kicker = 0;
        for (int r : ranks) if (r != p1 && r != p2) kicker = r;
        return {2, highPair, lowPair, kicker};
    }

    if (groups[0].first == 2) {
        int pair = groups[0].second;
        vector<int> kickers;
        for (int r : ranks) if (r != pair) kickers.push_back(r);
        sort(kickers.rbegin(), kickers.rend());
        return {1, pair, kickers[0], kickers[1], kickers[2]};
    }

    vector<int> v = {0};
    v.insert(v.end(), ranksDesc.begin(), ranksDesc.end());
    return v;
}

int main() {
    ifstream in("poker.txt");
    string c;
    int wins = 0;

    while (true) {
        vector<string> cards(10);
        for (int i = 0; i < 10; i++) {
            if (!(in >> cards[i])) {
                cout << wins << "\n"; // should be 376 on official file
                return 0;
            }
        }
        vector<string> h1(cards.begin(), cards.begin() + 5);
        vector<string> h2(cards.begin() + 5, cards.end());

        if (score(h1) > score(h2)) wins++;
    }
}

File: Python/Problem_054.py
from collections import Counter

RVAL = {**{str(i): i for i in range(2, 10)},
        "T": 10, "J": 11, "Q": 12, "K": 13, "A": 14}

def score(hand):
    # hand: list like ["5H","5C","6S","7S","KD"]
    ranks = sorted(RVAL[c[0]] for c in hand)
    suits = [c[1] for c in hand]

    is_flush = all(s == suits[0] for s in suits)

    # straight (handle A2345)
    is_straight = False
    straight_high = ranks[-1]
    if ranks == list(range(ranks[0], ranks[0] + 5)):
        is_straight = True
    elif ranks == [2, 3, 4, 5, 14]:
        is_straight = True
        straight_high = 5

    cnt = Counter(ranks)
    groups = sorted(((c, r) for r, c in cnt.items()), reverse=True)  # by count then rank

    ranks_desc = tuple(sorted(ranks, reverse=True))

    if is_flush and is_straight:
        if straight_high == 14:
            return (9,)  # royal flush
        return (8, straight_high)

    if groups[0][0] == 4:
        quad = groups[0][1]
        kicker = max(r for r in ranks if r != quad)
        return (7, quad, kicker)

    if groups[0][0] == 3 and groups[1][0] == 2:
        trip = groups[0][1]
        pair = groups[1][1]
        return (6, trip, pair)

    if is_flush:
        return (5,) + ranks_desc

    if is_straight:
        return (4, straight_high)

    if groups[0][0] == 3:
        trip = groups[0][1]
        kickers = tuple(sorted((r for r in ranks if r != trip), reverse=True))
        return (3, trip) + kickers

    if groups[0][0] == 2 and groups[1][0] == 2:
        pair1, pair2 = groups[0][1], groups[1][1]
        high_pair, low_pair = max(pair1, pair2), min(pair1, pair2)
        kicker = max(r for r in ranks if r != pair1 and r != pair2)
        return (2, high_pair, low_pair, kicker)

    if groups[0][0] == 2:
        pair = groups[0][1]
        kickers = tuple(sorted((r for r in ranks if r != pair), reverse=True))
        return (1, pair) + kickers

    return (0,) + ranks_desc


def euler_54(path="poker.txt") -> int:
    wins = 0
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            cards = line.strip().split()
            h1, h2 = cards[:5], cards[5:]
            if score(h1) > score(h2):
                wins += 1
    return wins


if __name__ == "__main__":
    print(euler_54())  # should print 376 with the official dataset
